programs/leap-util/actions/snapshot.cpp

```diff
@@ -10,29 +16,86 @@
+   to_json->add_option("--input-file,-i", opt->input_file, "Snapshot file to convert to json format, writes to <file>.json if output file not specified (tmp state dir used).")->required();
+   to_json->add_option("--output-file,-o", opt->output_file, "The file to write the output to (absolute or relative path).  If not specified then output is to <input-file>.json.");
 
-   sub->callback([this]() {
-      int rc = run_subcommand();
-      // properly return err code in main
-      if(rc) throw(CLI::RuntimeError(rc));
+   to_json->callback([this]() {
+      try {
+         int rc = run_subcommand();
+         if(rc) throw(CLI::RuntimeError(rc));
+      } catch(...) {
+         print_exception();
+         throw(CLI::RuntimeError(-1));
+      }
    });
 }
 
 int snapshot_actions::run_subcommand() {
    if(!opt->input_file.empty()) {
       if(!fc::exists(opt->input_file)) {
-         std::cerr << "cannot load snapshot, " << opt->input_file << " does not exist" << std::endl;
+         std::cerr << "cannot load snapshot, " << opt->input_file
+                   << " does not exist" << std::endl;
          return -1;
       }
    }
-   // todo: implementation here
 
+   bfs::path snapshot_path = opt->input_file;
+   bfs::path json_path = opt->output_file.empty()
+                               ? snapshot_path.generic_string() + ".json"
+                               : opt->output_file;
+   // pull chain id
+   auto infile = std::ifstream(snapshot_path.generic_string(),
+                               (std::ios::in | std::ios::binary));
+   istream_snapshot_reader reader(infile);
+   reader.validate();
+   auto chain_id = controller::extract_chain_id(reader);
+   infile.close();
+
+   // setup controller
+   bfs::path temp_dir = boost::filesystem::temp_directory_path() /
+                        boost::filesystem::unique_path();
+   bfs::path state_dir = temp_dir / "state";
+   bfs::path blocks_dir = temp_dir / "blocks";
+   std::unique_ptr<controller> control;
+   protocol_feature_set pfs;
+   controller::config cfg;
+   cfg.blocks_dir = blocks_dir;
+   cfg.state_dir = state_dir;
+
+   try {
+      auto infile = std::ifstream(snapshot_path.generic_string(),
+                                  (std::ios::in | std::ios::binary));
+      auto reader = std::make_shared<istream_snapshot_reader>(infile);
+      control.reset(new controller(cfg, std::move(pfs), chain_id));
+      control->add_indices();
+      control->startup([]() {}, []() { return false; }, reader);
+      infile.close();
+
+      ilog("Writing snapshot: ${s}", ("s", json_path.generic_string()));
+      auto snap_out = std::ofstream(json_path.generic_string(), (std::ios::out));
+      auto writer = std::make_shared<ostream_json_snapshot_writer>(snap_out);
+      control->write_snapshot(writer);
+      writer->finalize();
+      snap_out.flush();
+      snap_out.close();
+   } catch(const database_guard_exception& e) {
```

Comment:
Would be good to log that database size should be increased.

Reply:
done